### DISCUSSION TITLE
v4.2.0

### DISCUSSION THREAD

**Discussion by 4flixt** [do-mpc/do-mpc#169] [General]
# Major changes

## MX support
This addresses #34.
The do-mpc model  class can now be created with the ``symvar_type`` argument, defining whether the model is using CasADi ``SX`` or ``MX`` optimization variables.

```python
model = do_mpc.model.Model('continuous', 'MX')
```

all classes (MPC, MHE, Simulator ...) created from a MX model will also use MX variables.
From a users-perspective the change has no significant influence on the experience. 
It does, however, allow for significantly faster matrix vector operations, which is main motivation to use the ``MX`` support.

The new feature resulted in some major changes to the backend.
This is because CasADi does not allow (e.g.):
```python
x = MX.sym('x')
struct_symMX([
    entry('x', sym=x)
])
```
on which the model configuration heavily relied on.

 Most importantly:
- The Model class attributes ``_x``, ``_u`` etc. are now dicts prior to calling ``Model.setup``.
- Calling ``model['x']`` still works prior to calling ``Model.setup`` but works differently internally
- a new method ``_convert2struct`` converts dicts (e.g. of all the states) to symbolic structures (used in ``Model.setup``). The only problem: These structs hold variables with the same name but which are different.
- a new method ``_substitute_struct_vars`` is introduced and substitutes the variables in the dicts in all expressions (e.g. ``_rhs`` with the new variables from the symbolic structs.
- the MHE also required some major internal changes. The problem is that we split the parameters (``_p``) for the MHE into estimated and set parameters. Splitting symbolic variables with the MX type is problematic.

# Minor changes
- Solved #149 : Option to only have a single slack variable (for each soft-constraint) over the entire horizon

# Bug fixes
- Resolves #89. Discrete-time model now inherits its properties to MHE/MPC etc.

<hr /><em>This discussion was created from the release <a href='https://github.com/do-mpc/do-mpc/releases/tag/v4.2.0'>v4.2.0</a>.</em>

**Comment by 4flixt** [comment]
We are looking forward for your feedback on our latest release :)

**Reply by cahly** [reply]
This is great news! Will check it out!
